server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/EventDeadLettersIntegrationTest.java

```diff
@@ -0,0 +1,464 @@
+            if (retries.containsKey(event.getEventId())) {
+                currentRetries = retries.get(event.getEventId());
+            } else {
+                failedEvents.add(event);
+            }
+
+            if (currentRetries < retriesBeforeSuccess) {
+                retries.put(event.getEventId(), currentRetries + 1);
+                throw new RuntimeException("throw to trigger retry");
+            } else {
+                failedEvents.remove(event);
+                retries.remove(event.getEventId());
+                successfulEvents.add(event);
+            }
+        }
+
+        List<Event> getFailedEvents() {
+            return failedEvents;
+        }
+
+        List<Event> getSuccessfulEvents() {
+            return successfulEvents;
+        }
+
+        void setRetriesBeforeSuccess(int retriesBeforeSuccess) {
+            this.retriesBeforeSuccess = retriesBeforeSuccess;
+        }
+
+        void clear() {
+            init();
+        }
+    }
+
+    private static final String DOMAIN = "domain.tld";
+    private static final String BOB = "bob@" + DOMAIN;
+    private static final String BOB_PASSWORD = "bobPassword";
+    private static final String EVENTS_ACTION = "reDeliver";
+    private static final String SERIALIZED_GROUP = new RetryEventsListenerGroup().asString();
+
+    RetryEventsListener retryEventsListener = new RetryEventsListener();
+
+    @ClassRule
+    public static DockerCassandraRule cassandra = new DockerCassandraRule();
+
+    @Rule
+    public CassandraJmapTestRule cassandraJmapTestRule = CassandraJmapTestRule.defaultTestRule();
+
+    private GuiceJamesServer guiceJamesServer;
+    private MailboxProbe mailboxProbe;
+
+    private EventSerializer eventSerializer = new EventSerializer(new InMemoryId.Factory(), new InMemoryMessageId.Factory());
+
+    @Before
+    public void setUp() throws Exception {
+        retryEventsListener.clear();
+        guiceJamesServer = cassandraJmapTestRule.jmapServer(cassandra.getModule())
+            .overrideWith(new WebAdminConfigurationModule())
+            .overrideWith(binder -> Multibinder.newSetBinder(binder, MailboxListener.GroupMailboxListener.class).addBinding().toInstance(retryEventsListener));
+        guiceJamesServer.start();
+
+        DataProbe dataProbe = guiceJamesServer.getProbe(DataProbeImpl.class);
+        mailboxProbe = guiceJamesServer.getProbe(MailboxProbeImpl.class);
+        WebAdminGuiceProbe webAdminGuiceProbe = guiceJamesServer.getProbe(WebAdminGuiceProbe.class);
+
+        RestAssured.requestSpecification = WebAdminUtils.buildRequestSpecification(webAdminGuiceProbe.getWebAdminPort())
+            .build();
+
+        dataProbe.addDomain(DOMAIN);
+        dataProbe.addUser(BOB, BOB_PASSWORD);
+    }
+
+    @After
+    public void tearDown() {
+        guiceJamesServer.stop();
+    }
+
+    @Test
+    public void failedEventShouldBeStoredInRetryListenerFailedEventsList() {
+        retryEventsListener.setRetriesBeforeSuccess(4);
```

Comment:
Might be better to extract those numbers into constants...

Reply:
Didn't, just not sure that this is really relevant. WDYT?

Reply:
if you put 3, it's the max retires for success by default. So up to 3 retries it's a success. We put 4 in this case to exceed the number of successful retries and fail the delivery of the event intentionally, so it can be stored in dead letters

Reply:
Yes, I know it. But not sure for other member know exactly what `4` is. But OK. Never mind.